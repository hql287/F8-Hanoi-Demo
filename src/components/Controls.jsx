// React Libraries
import path from 'path';
import React, {Component} from 'react';
import styled from 'styled-components';

export const Dashboard = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  border-radius: 10px;
  padding: 20px;
  button {
    margin: 0 5px;
  }
`;

export const Button = styled.button`

`;

class Controls extends Component {
  render() {
    const { jump, moveLeft, moveRight, toggleLabels } = this.props;
    return (
      <Dashboard>
        <Button onClick={moveLeft}>← Move Left</Button>
        <Button onClick={jump}>↑ Jump</Button>
        <Button onClick={moveRight}>Move Right →</Button>
        <Button onClick={toggleLabels}>Toogle Labels</Button>
      </Dashboard>
    );
  }
}

export default Controls;

