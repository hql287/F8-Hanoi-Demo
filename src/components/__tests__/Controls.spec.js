// Libs
import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';

// Component to test
import Controls, {Dashboard, Button} from '../Controls';

// Mock
const moveLeft     = jest.fn();
const moveRight    = jest.fn();
const jump         = jest.fn();
const toggleLabels = jest.fn();

// Setup
describe('Renders the Controls component correctly', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Controls
        jump={jump}
        moveLeft={moveLeft}
        moveRight={moveRight}
        toggleLabels={toggleLabels}
      />,
    );
  });

  it('should render a Dashboard component', () => {
    expect(wrapper.find(Dashboard)).toHaveLength(1);
    expect(wrapper.find(Dashboard)).not.toHaveLength(2);
  });

  it('should render 4 Buttons', () => {
    expect(wrapper.find(Button)).toHaveLength(4);
    expect(wrapper.find(Button)).not.toHaveLength(3);
  });


  // Test style Rules
  it('should render components with correct styles', () => {

  });

  // Snapshot Testing
  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <Controls
          jump={jump}
          moveLeft={moveLeft}
          moveRight={moveRight}
          toggleLabels={toggleLabels}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('it should handle clicks correctly', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Controls
        jump={jump}
        moveLeft={moveLeft}
        moveRight={moveRight}
        toggleLabels={toggleLabels}
      />,
    );
  });

  it('handle moveLeft correctly', () => {
    const moveLeftBtn = wrapper.find(Button).first();
    moveLeftBtn.simulate('click');
    expect(moveLeft).toHaveBeenCalled();
    expect(moveLeft).toHaveBeenCalledTimes(1);
  });

  it('handle moveLeft correctly', () => {
    const moveRightBtn = wrapper.find(Button).at(2);
    moveRightBtn.simulate('click');
    expect(moveRight).toHaveBeenCalled();
    expect(moveRight).toHaveBeenCalledTimes(1);
  });

  it('handle Jump correctly', () => {
    const jumpBtn = wrapper.find(Button).at(1);
    jumpBtn.simulate('click');
    expect(jump).toHaveBeenCalled();
    expect(jump).toHaveBeenCalledTimes(1);
  });

  it('handle toggleLabels correctly', () => {
    const toggleLabelsBtn = wrapper.find(Button).last();
    toggleLabelsBtn.simulate('click');
    expect(toggleLabels).toHaveBeenCalled();
    expect(toggleLabels).toHaveBeenCalledTimes(1);
  });
})
