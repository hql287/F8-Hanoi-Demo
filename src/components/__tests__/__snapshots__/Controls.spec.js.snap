// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the Controls component correctly matches snapshot 1`] = `
<div
  className="Controls__Dashboard-cQakuF cCtKla"
>
  <button
    className="Controls__Button-gzvyre dvIyxS"
    onClick={[MockFunction]}
  >
    ← Move Left
  </button>
  <button
    className="Controls__Button-gzvyre dvIyxS"
    onClick={[MockFunction]}
  >
    ↑ Jump
  </button>
  <button
    className="Controls__Button-gzvyre dvIyxS"
    onClick={[MockFunction]}
  >
    Move Right →
  </button>
  <button
    className="Controls__Button-gzvyre dvIyxS"
    onClick={[MockFunction]}
  >
    Toogle Labels
  </button>
</div>
`;
